bin/yarn2nix.js

```diff
@@ -2,45 +2,96 @@
 "use strict";
 
 const path = require("path");
+const cp = require("child_process");
 
-const HEAD = `
-{fetchurl, linkFarm}: rec {
-  offline_cache = linkFarm "offline" packages;
-  packages = [
-`.trim();
+const remoteToString = function(remote) {
+  return `
+    {
+      name = "${remote.file_name}";
+      path = fetchurl {
+        name = "${remote.file_name}";
+        url  = "${remote.url}";
+        sha1 = "${remote.sha1}";
+      };
+    }`;
+}
 
-function generateNix(lockedDependencies) {
-  let found = {};
+const localToString = function(local) {
+  return `
+    {
+      name = "${local.file_name}";
+      path = ${local.file_name};
+    }`;
+}
 
-  console.log(HEAD)
+const arrayToString = function(elements) {
+  return `[
+    ${elements.join("\n")}
+  ]`;
+}
 
-  for (var depRange in lockedDependencies) {
-    let dep = lockedDependencies[depRange];
+const astToString = function(ast) {
+  const remotes = ast.links
+    .filter(function(link) { return link.hasOwnProperty('remote'); })
+    .map(function(link) { return remoteToString(link.remote); });
 
-    let depRangeParts = depRange.split('@');
-    let [url, sha1] = dep["resolved"].split("#");
-    let file_name = path.basename(url)
+  const locals = ast.links
+    .filter(function(link) { return link.hasOwnProperty('local'); })
+    .map(function(link) { return localToString(link.local); });
 
-    if (found.hasOwnProperty(file_name)) {
-      continue;
-    } else {
-      found[file_name] = null;
-    }
+  return `
+  {fetchurl, linkFarm}: rec {
+    offline_cache = linkFarm "offline" packages;
+    packages = ${arrayToString(remotes)};
+    localPackages = ${arrayToString(locals)};
+  }`;
+};
 
+function yarnToAst(lockedDependencies) {
+  let found = {};
 
-    console.log(`
-    {
-      name = "${file_name}";
-      path = fetchurl {
-        name = "${file_name}";
-        url  = "${url}";
-        sha1 = "${sha1}";
-      };
-    }`)
-  }
+  return {
+    links: Object.keys(lockedDependencies).map(function(depRange) {
+      let dep = lockedDependencies[depRange];
+
+      let depRangeParts = depRange.split('@');
+      if ('resolved' in dep) {
+        let url;
+        let sha1;
+        let file_name;
+        if (/codeload.github.com.*tar.gz\//.test(dep["resolved"])) {
+          url = dep["resolved"];
+          sha1 = cp.execSync("curl -sS " + url + " | shasum | cut -d \" \" -f 1").toString().trim();
```

Comment:
This can probably be done in Node.js without shelling out.

Reply:
Yes please :)

Reply:
Actually, the whole point of yarn2nix is that it doesn't need network access to build the nix file. Because it's pure, it can be loaded recursively.

I found that patching the yarn.lock to add the `#sha1` to the github urls works fine (but is an operation that has to be done by hand)

Reply:
This command converts a git commit into a tarball SHA, which is pure barring the same URL responding with different tarballs at different times. Do you foresee any issues with this?

Maybe we could automate the process of appending `#sha1` and call the script `hashify-urls.js` or something.

Reply:
>barring the same URL responding with different tarballs at different times

I think this does happen — this is part of why nix and nixpkgs have fetchTarball which unpacks it after fetching, so that different but equivalent tarballs have the same nar hash.